structopt-derive/src/lib.rs

```diff
@@ -141,6 +151,8 @@ fn gen_constructor(fields: &Punctuated<Field, Comma>) -> quote::Tokens {
                 _ => quote!( .unwrap() )
             };
             quote!(#field_name: #subcmd_type::from_subcommand(matches.subcommand())#unwrapper)
+        } else if attrs.should_flatten() {
+            quote!(#field_name: StructOpt::from_clap(matches))
```

Comment:
The usage seems always to be `if attrs.is_subcommand() { ... } else if attrs.should_flatten() { ... } else { ... }`.

Maybe an `enum Mode { SubCommand, Flatten, Arg }` (or other naming if you're more inspired than me) would be better?

Also, checking that we don't `subcommand` and `flatten` on the same field seems reasonable.